server/src/main/java/org/elasticsearch/search/SearchUtils.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search;
+
+import org.apache.lucene.search.IndexSearcher;
+import org.elasticsearch.monitor.jvm.JvmStats;
+import org.elasticsearch.threadpool.ThreadPool;
+
+public final class SearchUtils {
+
+    /**
+     * Configure the maximum lucene query clause count based on the size of the search thread pool and maximum heap
+     */
+    public static void configureMaxClauses(ThreadPool threadPool) {
+        int searchThreadPoolSize = threadPool.info(ThreadPool.Names.SEARCH).getMax();
+        long heapSize = JvmStats.jvmStats().getMem().getHeapMax().getGb();
+        configureMaxClauses(searchThreadPoolSize, heapSize);
```

Comment:
What happens if you start Elasticsearch with 512MB of heap, does it mean you get zero max clauses?

Reply:
If `heapSize` is zero then we stick with the default value from lucene, currently 1024.  But maybe we should stick with the current ES default of 4096 instead?

Reply:
Have updated this so that we will always return a minimum of 4096, or larger if the heap/cpu size permits.

Reply:
Would this work if we used the heap size in GB as a fractional number instead?